README.md
# Bows
![Colors](https://raw.github.com/latentflip/bows/master/example/cols.png)
Safe, production happy, colourful logging for chrome - makes reading your logs _much_ easier.


(Rain)bows makes logging debug messages in your apps much nicer.
- It allows you to create custom loggers for each module in your app, that prefix all log messages with the name of the app, so that you can scan the messages more easily.
- It colors the prefix differently and distinctly for each logger/module so that it's even easier to read.
- It can be safely used in production, where logging will be disabled by default, so that you can leave log messages in your code.
- Loggers safely wrap console.log, to maintain the line number from where they are called in the console output.

![Example Output](https://raw.github.com/latentflip/bows/master/example/realexample.png)

## Installation.

If you are using browserify, you'll want something like:

```
npm install bows --save
```

If you aren't using browserify, download either [bows.js](https://raw.github.com/latentflip/bows/master/dist/bows.js) or [bows.min.js](https://raw.github.com/latentflip/bows/master/dist/bows.min.js).

## Features

* Easily create prefixes for your logs, so that you can distinguish between logs from different parts of your app easily.
* In chrome prefixes will be color coded for even easier identification.
* Can be safely used in production, as logs will be disabled for your users, but can be enabled by you with a local storage flag.
* Greppable logs by setting `localStorage.debug = /Foo/` to only display logs for modules matching the regex to help you focus in development.


## Usage
- Works great in browserify and the browser.
- Creating a new logger:
  - Browserify: `var log = require('bows')("My Module Name")`
  - Browser: `var log = bows("My Module Name")`
- Then using it is easy:
  - `log("Module loaded") //-> "My Module Name | Module Loaded"`
  - `log("Did something") //-> "My Module Name | Did something"`
- Typically each seperate module/view/etc in your app would create it's own logger. It will be assigned it's own color to make it easy to spot logs from different modules.
- Logging is disabled by default. To enable logging, set `localStorage.debug = true` in your console and refresh the page.
- You can leave the code in in production, and log() will just safely no-op unless localStorage.debug is set.

## Example

```javascript
  //Should be set in your console to see messages
  localStorage.debug = true
  //Configure the max length of module names (optional)
  bows.config({ padLength: 10 })

  var logger1 = bows('Module 1')
  var logger2 = bows('Module 2')
  var logger3 = bows('Module 3')

  logger1("We started up")
  logger2("We did something too")
  logger3("I'm here")
  logger3("I'm still here")
  logger2("I'm tired")
  logger1("We're done here")
```

Result:

![Example Output](https://raw.github.com/latentflip/bows/master/example/example.png)

## License & Credits

MIT

Copyright [@Philip\_Roberts](http://twitter.com/philip\_roberts) [latentflip.com](http://latentflip.com).

Bows depends on [andlog](http://github.com/henrikjoreteg/andlog), a nice little logging module by [@HenrikJoreteg](https://twitter.com/henrikjoreteg).


## Contributing

Please feel free to raise issues, or make contributions:

```bash
git clone https://github.com/latentflip/bows.git
cd bows
npm install #install dependencies
#edit bows.js
node build.js #build dist/bows.js and dist/bows.min.js
```



CHANGELOG.md
## November 13, 2013 (v0.3)
* Add support for log, debug, error, warn, info as methods on the logger: 

    ```
    logger = bows("MyModule")
    logger("This logs")
    logger.log("This logs")
    logger.warn("This warns")
    logger.error("This errors")
    logger.debug("This debugs")
    logger.info("This infos")
    ```

* Fix bug where the padLength config wasn't doing anything
* Add support for colors in firebug
* Disable color string garbage where colors are not supported


test.js
var assert = require('assert'),
    bows = require('./bows.js');

var foo, bar, cat;
var fooPrefix, barPrefix, catPrefix;


foo = bows('Foo')
fooPrefix = foo();

assert.equal(fooPrefix[0], '%cFoo |', 'Foo prefix set');
assert.equal(fooPrefix[1], 'color: hsl(0,50%,50%); font-weight: bold', 'Foo color set to black');

// Add more
bar = bows('Barrr');
barPrefix = bar();
fooPrefix = foo();

assert.equal(barPrefix[0], '%cBarrr |', 'Bar prefix set');
assert.equal(barPrefix[1], 'color: hsl(180,50%,50%); font-weight: bold', 'Bar color set to the next one');

assert.equal(fooPrefix[0], '%cFoo   |', 'Foo prefix length updated');
assert.equal(fooPrefix[1], 'color: hsl(0,50%,50%); font-weight: bold', 'Foo color not updated');



//Add even more

var cat = bows('Cat');

catPrefix = cat();
barPrefix = bar();
fooPrefix = foo();

assert.equal(catPrefix[0], '%cCat   |', 'Bar prefix set');
assert.equal(catPrefix[1], 'color: hsl(240,50%,50%); font-weight: bold', 'Cat color set to the next one');

assert.equal(barPrefix[0], '%cBarrr |', 'Bar prefix set');
assert.equal(barPrefix[1], 'color: hsl(120,50%,50%); font-weight: bold', 'Bar color updated');

assert.equal(fooPrefix[0], '%cFoo   |', 'Foo prefix length updated');
assert.equal(fooPrefix[1], 'color: hsl(0,50%,50%); font-weight: bold', 'Foo color not updated');

bows.color = false;
assert.equal(foo('a log string')[0], 'a log string', 'disable logging');

console.log('All tests pass');


bows.bundle.js
(function(e){if("function"==typeof bootstrap)bootstrap("bows",e);else if("object"==typeof exports)module.exports=e();else if("function"==typeof define&&define.amd)define(e);else if("undefined"!=typeof ses){if(!ses.ok())return;ses.makeBows=e}else"undefined"!=typeof window?window.bows=e():global.bows=e()})(function(){var define,ses,bootstrap,module,exports;
return (function(e,t,n){function i(n,s){if(!t[n]){if(!e[n]){var o=typeof require=="function"&&require;if(!s&&o)return o(n,!0);if(r)return r(n,!0);throw new Error("Cannot find module '"+n+"'")}var u=t[n]={exports:{}};e[n][0].call(u.exports,function(t){var r=e[n][1][t];return i(r?r:t)},u,u.exports)}return t[n].exports}var r=typeof require=="function"&&require;for(var s=0;s<n.length;s++)i(n[s]);return i})({1:[function(require,module,exports){
(function() {
  var bows, goldenRatio, hue, logger, padLength, yieldColor;

  logger = require('andlog');

  goldenRatio = 0.618033988749895;
  hue = 0;

  yieldColor = function() {
    hue += goldenRatio;
    hue = hue % 1;
    return hue * 360;
  };
  window.yieldColor = yieldColor;

  padLength = 10;

  bows = function(str) {
    var msg;
    msg = "%c" + (str.slice(0, padLength));
    msg += Array(padLength + 3 - msg.length).join(' ') + '|';

    return logger.log.bind(logger, msg, "color: hsl(" + (yieldColor()) + ",99%,40%); font-weight: bold");
  };

  bows.config = function(config) {
    if (config.padLength) {
      return padLength = config.padLength;
    }
  };

  if (typeof module !== 'undefined') {
    module.exports = bows;
  } else {
    window.bows = bows;
  }
}).call(this);

},{"andlog":2}],2:[function(require,module,exports){
(function(global){// follow @HenrikJoreteg and @andyet if you like this ;)
(function (window) {
    var ls = window.localStorage,
        out = {},
        inNode = (
          typeof global !== 'undefined' &&
          (global !== window) &&
          typeof window.global === 'undefined'
        )

    if (inNode) {
        module.exports = console;
        return;
    }

    if (ls && ls.debug && window.console) {
        out = window.console;
    } else {
        var methods = "assert,count,debug,dir,dirxml,error,exception,group,groupCollapsed,groupEnd,info,log,markTimeline,profile,profileEnd,time,timeEnd,trace,warn".split(","),
            l = methods.length,
            fn = function () {};
        
        while (l--) {
            out[methods[l]] = fn;
        }
    }
    if (typeof exports !== 'undefined') {
        module.exports = out;
    } else {
        window.console = out;
    }
})(this);

})(self)
},{}]},{},[1])(1)
});
;

bows.js
(function() {
  function checkColorSupport() {
    var chrome = !!window.chrome,
        firefox = /firefox/i.test(navigator.userAgent),
        firebug = firefox && !!window.console.exception;

    return chrome || firebug;
  }

  var inNode = typeof window === 'undefined',
      ls = !inNode && window.localStorage,
      debug = ls.debug,
      logger = require('andlog'),
      hue = 0,
      padLength = 15,
      noop = function() {},
      colorsSupported = ls.debugColors || checkColorSupport(),
      yieldColor,
      bows,
      debugRegex;

  yieldColor = function() {
    var goldenRatio = 0.618033988749895;
    hue += goldenRatio;
    hue = hue % 1;
    return hue * 360;
  };

  debugRegex = debug && debug[0]==='/' && new RegExp(debug.substring(1,debug.length-1));

  bows = function(str) {
    var msg, colorString, logfn;
    msg = (str.slice(0, padLength));
    msg += Array(padLength + 3 - msg.length).join(' ') + '|';

    if (debugRegex && !str.match(debugRegex)) return noop;

    if (colorsSupported) {
      var color = yieldColor();
      msg = "%c" + msg;
      colorString = "color: hsl(" + (color) + ",99%,40%); font-weight: bold";

      logfn = logger.log.bind(logger, msg, colorString);
      ['log', 'debug', 'warn', 'error', 'info'].forEach(function (f) {
        logfn[f] = logger[f].bind(logger, msg, colorString);
      });
    } else {
      logfn = logger.log.bind(logger, msg);
      ['log', 'debug', 'warn', 'error', 'info'].forEach(function (f) {
        logfn[f] = logger[f].bind(logger, msg);
      });
    }

    return logfn;
  };

  bows.config = function(config) {
    if (config.padLength) {
      this.padLength = config.padLength;
    }
  };

  if (typeof module !== 'undefined') {
    module.exports = bows;
  } else {
    window.bows = bows;
  }
}).call();
